mllib/src/main/scala/org/apache/spark/mllib/tree/GradientBoostedTrees.scala

```diff
@@ -177,102 +177,108 @@ object GradientBoostedTrees extends Logging {
     treeStrategy.assertValid()
 
     // Cache input
-    if (input.getStorageLevel == StorageLevel.NONE) {
+    val persistedInput = if (input.getStorageLevel == StorageLevel.NONE) {
       input.persist(StorageLevel.MEMORY_AND_DISK)
-    }
+      true
+    } else false
 
-    timer.stop("init")
+    try {
```

Comment:
I override and intercept most RDD calls in my own code to optimize them (which is how I found this in the first place) and we use try-with-resource to persist and unpersist. It's too bad you cant do that simply in scala.

If you want the simpler solution let me know. I can change it.